dask/array/routines.py

```diff
@@ -160,6 +160,59 @@ def dot(a, b):
     return tensordot(a, b, axes=((a.ndim - 1,), (b.ndim - 2,)))
 
 
+def _func1d_fmt(arr,
+                func1d,
+                func1d_args,
+                func1d_kwargs,
+                ndim_before=0,
+                ndim_after=0):
+    # Drop singleton dimensions from map_blocks to get 1D array.
+    arr1d = arr.flatten()
+
+    # Compute result and ensure it is an array (consistent with NumPy).
+    res = np.asarray(func1d(arr1d, *func1d_args, **func1d_kwargs))
+
+    # Tack on singleton dimensions that map_blocks will expect.
+    res = res[ndim_before * (None,) + (Ellipsis,) + ndim_after * (None,)]
+
+    return res
+
+
+@wraps(np.apply_along_axis)
+def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+    arr = asarray(arr)
+
+    # Validate and normalize axis.
+    arr.shape[axis]
+    axis = len(arr.shape[:axis])
+
+    # Test out some data with the function.
+    test_data = np.ones((1,), dtype=arr.dtype)
+    test_result = np.array(func1d(test_data))
+
+    # Rechunk so that func1d is applied over the full axis.
+    arr = arr.rechunk(
+        axis * (1,) + arr.shape[axis:axis + 1] + (arr.ndim - axis) * (1,)
+    )
+
+    # Map func1d over the data to get the result
+    # Adds other axes as needed.
+    result = arr.map_blocks(
+        _func1d_fmt,
+        dtype=test_result.dtype,
+        chunks=(arr.chunks[:axis] + test_result.shape + arr.chunks[axis + 1:]),
```

Comment:
Guessing that the shape of the result is not somehow dependent on the shape of the data. Could fix this by passing a 1D array of the expected shape along. Could be painful if the computation is slow (note this would be done during graph setup). Though am hoping that this is not typically necessary. Would be curious to hear other thoughts on this.